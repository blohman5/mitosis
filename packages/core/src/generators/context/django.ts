import { getContextWithSymbolKey } from './helpers/context-with-symbol-key';

export const contextToDjango = getContextWithSymbolKey;
