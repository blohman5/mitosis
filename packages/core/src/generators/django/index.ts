export * from './django';
export * from './types';
