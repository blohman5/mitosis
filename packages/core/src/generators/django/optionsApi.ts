import { getComponentsUsed } from '@/helpers/get-components-used';
import { getCustomImports } from '@/helpers/get-custom-imports';
import { getStateObjectStringFromComponent } from '@/helpers/get-state-object-string';
import { checkIsDefined } from '@/helpers/nullable';
import { checkIsComponentImport } from '@/helpers/render-imports';
import { BaseHook, MitosisComponent } from '@/types/mitosis-component';
import json5 from 'json5';
import { kebabCase, uniq } from 'lodash';
import { encodeQuotes, getContextKey, getContextValue } from './helpers';
import { DjangoDefaultProps, DjangoPropsDefinition, ToDjangoOptions } from './types';

const getContextProvideString = (json: MitosisComponent, options: ToDjangoOptions) => {
  return `{
    ${Object.values(json.context.set)
      .map((setVal) => {
        const key = getContextKey(setVal);
        return `[${key}]: ${getContextValue(setVal)}`;
      })
      .join(',')}
  }`;
};

function getContextInjectString(component: MitosisComponent, options: ToDjangoOptions) {
  let str = '{';

  const contextGetters = component.context.get;

  for (const key in contextGetters) {
    const context = contextGetters[key];
    str += `
      ${key}: ${encodeQuotes(getContextKey(context))},
    `;
  }

  str += '}';
  return str;
}

const generateComponentImport =
  (options: ToDjangoOptions) =>
  (componentName: string): string => {
    if (options.asyncComponentImports) {
      return `'${componentName}': defineAsyncComponent(${componentName})`;
    } else {
      return `'${componentName}': ${componentName}`;
    }
  };

const generateComponents = (componentsUsed: string[], options: ToDjangoOptions): string => {
  if (componentsUsed.length === 0) {
    return '';
  } else {
    return `components: { ${componentsUsed.map(generateComponentImport(options)).join(',')} },`;
  }
};

const appendToDataString = ({
  dataString,
  newContent,
}: {
  dataString: string;
  newContent: string;
}) => dataString.replace(/}$/, `${newContent}}`);

export function generateOptionsApiScript(
  component: MitosisComponent,
  options: ToDjangoOptions,
  path: string | undefined,
  template: string,
  props: string[],
  onUpdateWithDeps: BaseHook[],
  onUpdateWithoutDeps: BaseHook[],
) {
  const { exports: localExports } = component;
  const localVarAsData: string[] = [];
  const localVarAsFunc: string[] = [];
  const isTs = options.typescript;
  if (localExports) {
    Object.keys(localExports).forEach((key) => {
      if (localExports[key].usedInLocal) {
        if (localExports[key].isFunction) {
          localVarAsFunc.push(key);
        } else {
          localVarAsData.push(key);
        }
      }
    });
  }

  let dataString = getStateObjectStringFromComponent(component, {
    data: true,
    functions: false,
    getters: false,
  });

  // Append refs to data as { foo, bar, etc }
  dataString = appendToDataString({
    dataString,
    newContent: getCustomImports(component).join(','),
  });

  if (localVarAsData.length) {
    dataString = appendToDataString({ dataString, newContent: localVarAsData.join(',') });
  }

  const getterString = getStateObjectStringFromComponent(component, {
    data: false,
    getters: true,
    functions: false,
  });

  let functionsString = getStateObjectStringFromComponent(component, {
    data: false,
    getters: false,
    functions: true,
  });

  const includeClassMapHelper = template.includes('_classStringToObject');

  if (includeClassMapHelper) {
    functionsString = functionsString.replace(
      /}\s*$/,
      `_classStringToObject(str${isTs ? ': string' : ''}) {
        const obj${isTs ? ': Record<string, boolean>' : ''} = {};
        if (typeof str !== 'string') { return obj }
        const classNames = str.trim().split(/\\s+/);
        for (const name of classNames) {
          obj[name] = true;
        }
        return obj;
      }  }`,
    );
  }

  if (localVarAsFunc.length) {
    functionsString = functionsString.replace(/}\s*$/, `${localVarAsFunc.join(',')}}`);
  }

  const getCompositionPropDefinition = ({
    options,
    component,
    props,
  }: {
    options: ToDjangoOptions;
    component: MitosisComponent;
    props: string[];
  }) => {
    const isTs = options.typescript;
    let str = ` def get_context_data(`;
    str += `${json5.stringify(props)}):
    `;
    str += `return {
    }`;
    return str;
  };

  // Component references to include in `component: { YourComponent, ... }
  const componentsUsedInTemplate = Array.from(getComponentsUsed(component))
    .filter((name) => name.length && !name.includes('.') && name[0].toUpperCase() === name[0])
    // Strip out components that compile away
    .filter((name) => !['For', 'Show', 'Fragment', 'Slot', component.name].includes(name));

  // get default imports from component files
  const importedComponents = component.imports
    .filter(checkIsComponentImport)
    .map((imp) => Object.entries(imp.imports).find(([_, value]) => value === 'default')?.[0])
    .filter(checkIsDefined);

  const componentsUsed = uniq([...componentsUsedInTemplate, ...importedComponents]);

  const getPropDefinition = ({
    component,
    props,
  }: {
    component: MitosisComponent;
    props: string[];
  }) => {
    const propsDefinition: DjangoPropsDefinition<DjangoDefaultProps> = Array.from(props).filter(
      (prop) => prop !== 'children' && prop !== 'class',
    );
    let str = 'props: ';

    if (component.defaultProps) {
      const defalutPropsString = propsDefinition
        .map((prop) => {
          const value = component.defaultProps!.hasOwnProperty(prop)
            ? component.defaultProps![prop]?.code
            : 'undefined';
          return `${prop}: { default: ${value} }`;
        })
        .join(',');

      str += `{${defalutPropsString}}`;
    } else {
      str += `${json5.stringify(propsDefinition)}`;
    }
    return `${str},`;
  };

  return `
  ${
    getterString.length < 4
      ? ''
      : `
  computed: ${getterString},
  `
  }
  @component.register("${
    !component.name
      ? ''
      : `${path && options.namePrefix?.(path) ? options.namePrefix?.(path) + '-' : ''}${kebabCase(
          component.name,
        )}`
  }")
class ${
    !component.name
      ? ''
      : `${path && options.namePrefix?.(path) ? options.namePrefix?.(path) + '-' : ''}${kebabCase(
          component.name,
        )}`
  }(component.Component):
  ${props.length ? getCompositionPropDefinition({ component, props, options }) : ''}
`;
}
