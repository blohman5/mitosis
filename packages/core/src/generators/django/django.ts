import { convertTypeScriptToJS } from '@/helpers/babel-transform';
import { createSingleBinding } from '@/helpers/bindings';
import { dedent } from '@/helpers/dedent';
import { fastClone } from '@/helpers/fast-clone';
import { getProps } from '@/helpers/get-props';
import { isMitosisNode } from '@/helpers/is-mitosis-node';
import { mapRefs } from '@/helpers/map-refs';
import { initializeOptions } from '@/helpers/merge-options';
import { processOnEventHooksPlugin } from '@/helpers/on-event';
import { CODE_PROCESSOR_PLUGIN } from '@/helpers/plugins/process-code';
import { processHttpRequests } from '@/helpers/process-http-requests';
import { replaceStateIdentifier } from '@/helpers/replace-identifiers';
import { isSlotProperty } from '@/helpers/slots';
import { stripMetaProperties } from '@/helpers/strip-meta-properties';
import { collectCss } from '@/helpers/styles/collect-css';
import { MitosisComponent } from '@/types/mitosis-component';
import { TranspilerGenerator } from '@/types/transpiler';
import { flow } from 'fp-ts/lib/function';
import { pickBy, size } from 'lodash';
import { format } from 'prettier/standalone';
import traverse from 'traverse';
import {
  Plugin,
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../../modules/plugins';
import { FUNCTION_HACK_PLUGIN } from '../helpers/functions';
import { blockToDjango } from './blocks';
import { getOnUpdateHookName, processBinding, renameMitosisComponentsToKebabCase } from './helpers';
import { generateOptionsApiScript } from './optionsApi';
import { ToDjangoOptions } from './types';

// Transform <foo.bar key="value" /> to <component :is="foo.bar" key="value" />
function processDynamicComponents(json: MitosisComponent, _options: ToDjangoOptions) {
  traverse(json).forEach((node) => {
    if (isMitosisNode(node)) {
      if (node.name.includes('.')) {
        node.bindings.is = createSingleBinding({ code: node.name });
        node.name = 'component';
      }
    }
  });
}

function processForKeys(json: MitosisComponent, _options: ToDjangoOptions) {
  traverse(json).forEach((node) => {
    if (isMitosisNode(node)) {
      if (node.name === 'For') {
        const firstChild = node.children[0];
        if (firstChild && firstChild.bindings.key) {
          node.bindings.key = firstChild.bindings.key;
          delete firstChild.bindings.key;
        }
      }
    }
  });
}

/**
 * This plugin handle `onUpdate` code that watches dependencies.
 * We need to apply this workaround to be able to watch specific dependencies in Vue 2: https://stackoverflow.com/a/45853349
 *
 * We add a `computed` property for the dependencies, and a matching `watch` function for the `onUpdate` code
 */
const onUpdatePlugin: Plugin = (options) => ({
  json: {
    post: (component) => {
      if (component.hooks.onUpdate) {
        component.hooks.onUpdate
          .filter((hook) => hook.deps?.length)
          .forEach((hook, index) => {
            const code = `get ${getOnUpdateHookName(index)} () {
            return {
              ${hook.deps
                ?.slice(1, -1)
                .split(',')
                .map((dep, k) => {
                  const val = dep.trim();
                  return `${k}: ${val}`;
                })
                .join(',')}
            }
          }`;

            component.state[getOnUpdateHookName(index)] = {
              code,
              type: 'getter',
            };
          });
      }
    },
  },
});

const BASE_OPTIONS: ToDjangoOptions = {
  api: 'options',
  defineComponent: true,
  casing: 'pascal',
};

export const componentToDjango: TranspilerGenerator<Partial<ToDjangoOptions>> =
  (userOptions) =>
  ({ component: _component, path }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain can be used
    let component = fastClone(_component);

    const options = initializeOptions({
      target: 'django',
      component,
      defaults: BASE_OPTIONS,
      userOptions: userOptions,
    });

    if (options.api === 'composition') {
      options.asyncComponentImports = false;
    }

    options.plugins.unshift(
      processOnEventHooksPlugin(),
      ...(options.api === 'options' ? [onUpdatePlugin] : []),
      ...(options.api === 'composition' ? [FUNCTION_HACK_PLUGIN] : []),
      CODE_PROCESSOR_PLUGIN((codeType) => {
        if (options.api === 'composition') {
          switch (codeType) {
            case 'hooks':
              return (code) => processBinding({ code, options, json: component });
            case 'state':
              return (code) => processBinding({ code, options, json: component });
            case 'bindings':
              return flow(
                // Strip types from any JS code that ends up in the template, because Vue does not support TS code in templates.
                convertTypeScriptToJS,
                (code) => processBinding({ code, options, json: component, codeType }),
              );
            case 'context-set':
              return (code) =>
                processBinding({ code, options, json: component, preserveGetter: true });
            case 'hooks-deps':
              return replaceStateIdentifier(null);
            case 'properties':
            case 'dynamic-jsx-elements':
            case 'types':
              return (c) => c;
          }
        } else {
          switch (codeType) {
            case 'hooks':
              return (code) => processBinding({ code, options, json: component });
            case 'bindings':
              return flow(
                // Strip types from any JS code that ends up in the template, because Vue does not support TS code in templates.
                convertTypeScriptToJS,
                (code) => processBinding({ code, options, json: component, codeType }),
              );
            case 'properties':
            case 'dynamic-jsx-elements':
            case 'hooks-deps':
            case 'types':
              return (c) => c;
            case 'state':
              return (c) => processBinding({ code: c, options, json: component });
            case 'context-set':
              return (code) =>
                processBinding({
                  code,
                  options,
                  json: component,
                  thisPrefix: '_this',
                  preserveGetter: true,
                });
          }
        }
      }),
    );

    processHttpRequests(component);
    processDynamicComponents(component, options);
    processForKeys(component, options);

    component = runPreJsonPlugins({ json: component, plugins: options.plugins });

    if (options.api === 'options') {
      mapRefs(component, (refName) => `this.$refs.${refName}`);
    }

    // need to run this before we process the component's code
    const props = Array.from(getProps(component));
    const elementProps = props.filter((prop) => !isSlotProperty(prop));
    const slotsProps = props.filter((prop) => isSlotProperty(prop));

    component = runPostJsonPlugins({ json: component, plugins: options.plugins });

    const css = collectCss(component, {
      prefix: options.cssNamespace?.() ?? undefined,
    });

    stripMetaProperties(component);

    const templateStrBody = component.children
      .map((item) => blockToDjango(item, options, { isRootNode: true }))
      .join('\n');

    const template =
      options.casing === 'kebab'
        ? renameMitosisComponentsToKebabCase(templateStrBody)
        : templateStrBody;

    const onUpdateWithDeps = component.hooks.onUpdate?.filter((hook) => hook.deps?.length) || [];
    const onUpdateWithoutDeps =
      component.hooks.onUpdate?.filter((hook) => !hook.deps?.length) || [];

    const getterKeys = Object.keys(pickBy(component.state, (i) => i?.type === 'getter'));

    // import from vue
    let djangoImports: string[] = [];
    if (options.asyncComponentImports) {
      djangoImports.push('defineAsyncComponent');
    }
    if (options.api === 'options' && options.defineComponent) {
      djangoImports.push('defineComponent');
    }
    if (options.api === 'composition') {
      onUpdateWithDeps.length && djangoImports.push('watch');
      component.hooks.onMount.length && djangoImports.push('onMounted');
      component.hooks.onUnMount?.code && djangoImports.push('onUnmounted');
      onUpdateWithoutDeps.length && djangoImports.push('onUpdated');
      size(getterKeys) && djangoImports.push('computed');
      size(component.context.set) && djangoImports.push('provide');
      size(component.context.get) && djangoImports.push('inject');
      size(
        Object.keys(component.state).filter((key) => component.state[key]?.type === 'property'),
      ) && djangoImports.push('ref');
      size(slotsProps) && djangoImports.push('useSlots');
    }

    const tsLangAttribute = options.typescript ? `lang='ts'` : '';
    //I question if this str is the main str everything is put into

    let str: string = dedent`
from django_components import component
from django_components import types as t

      ${(options.typescript && component.types?.join('\n')) || ''}

//within this is the component register stuff
//note how both things are from different packages
//For now I will remove one
      ${generateOptionsApiScript(
        component,
        options,
        path,
        template,
        elementProps,
        onUpdateWithDeps,
        onUpdateWithoutDeps,
      )}
    ${
      template.trim().length > 0
        ? `template: t.django_html = \"\"\"
      ${template}
    \"\"\"`
        : ''
    }

    ${
      !css.trim().length
        ? ''
        : `css: t.css = \"\"\"
      ${css}
    \"\"\"`
    }

    ${`js: t.js = \"\"\"
    \"\"\"`}
  `;

    str = runPreCodePlugins({
      json: component,
      code: str,
      plugins: options.plugins,
      options: { json: component },
    });
    if (true || options.prettier !== false) {
      try {
        str = format(str, {
          parser: 'django',
          plugins: [
            // To support running in browsers
            require('prettier/parser-typescript'),
            require('prettier/parser-html'),
            require('prettier/parser-postcss'),
            require('prettier/parser-babel'),
          ],
        });
      } catch (err) {
        console.warn('Could not prettify', { string: str }, err);
      }
    }
    str = runPostCodePlugins({ json: component, code: str, plugins: options.plugins });

    for (const pattern of removePatterns) {
      str = str.replace(pattern, '').trim();
    }
    str = str.replace(/<script(.*)>\n?<\/script>/g, '').trim();
    return str;
  };

// Remove unused artifacts like empty script or style tags
//Not needed for now
const removePatterns = [
  `<script2>
export default {};
</script>`,
  `<style44>
</style3>`,
];
