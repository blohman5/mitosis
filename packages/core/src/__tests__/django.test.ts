import { componentToDjango } from '../generators/django';
import { runTestsForTarget } from './test-generator';

describe('Django', () => {
  runTestsForTarget({
    target: 'django',
    generator: componentToDjango,
    options: {},
  });
});
