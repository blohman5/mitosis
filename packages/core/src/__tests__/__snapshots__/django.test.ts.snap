// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Django > svelte > Javascript Test > basic 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() name = \\"Steve\\";

  render() {
    return (
      <div>
        <input
          onInput={(event) => (this.name = event.target.value)}
          value={this.name}
        />
        Hello! I can run in React, Vue, Solid, or Liquid!
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > bindGroup 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() tortilla = \\"Plain\\";
  @State() fillings = [];

  render() {
    return (
      <div>
        <input
          type=\\"radio\\"
          value=\\"Plain\\"
          checked={this.tortilla === \\"Plain\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <input
          type=\\"radio\\"
          value=\\"Whole wheat\\"
          checked={this.tortilla === \\"Whole wheat\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <input
          type=\\"radio\\"
          value=\\"Spinach\\"
          checked={this.tortilla === \\"Spinach\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <br />
        <br />
        <input
          type=\\"checkbox\\"
          value=\\"Rice\\"
          checked={this.fillings === \\"Rice\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Beans\\"
          checked={this.fillings === \\"Beans\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Cheese\\"
          checked={this.fillings === \\"Cheese\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Guac (extra)\\"
          checked={this.fillings === \\"Guac (extra)\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <p>
          Tortilla:
          {this.tortilla}
        </p>
        <p>
          Fillings:
          {this.fillings}
        </p>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > bindProperty 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() value = \\"hello\\";

  render() {
    return <input value={this.value} />;
  }
}
"
`;

exports[`Django > svelte > Javascript Test > classDirective 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @Prop() disabled: any;

  @State() focus = true;

  render() {
    return (
      <input
        class={\`form-input \${props.disabled ? \\"disabled\\" : \\"\\"} \${
          state.focus ? \\"focus\\" : \\"\\"
        }\`}
      />
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > context 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() activeTab = 0;

  render() {
    return <div>{this.activeTab}</div>;
  }
}
"
`;

exports[`Django > svelte > Javascript Test > each 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() numbers = [\\"one\\", \\"two\\", \\"three\\"];

  render() {
    return (
      <ul>
        {this.numbers?.map((num) => (
          <li>{num}</li>
        ))}
      </ul>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > eventHandlers 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  log = function log(msg = \\"hello\\") {
    console.log(msg);
  };

  render() {
    return (
      <div>
        <button onClick={(a) => this.log(\\"hi\\")}>Log</button>
        <button onClick={(event) => this.log(event)}>Log</button>
        <button onClick={(event) => this.log(event)}>Log</button>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > html 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() html = \\"<b>bold</b>\\";

  render() {
    return <div innerHTML={this.html}></div>;
  }
}
"
`;

exports[`Django > svelte > Javascript Test > ifElse 1`] = `
"':' expected. (25:14)
  23 |       return (
  24 |
> 25 |         {this.show ? (
     |              ^
  26 |     <button  onClick={event => this.toggle(event)} > Hide </button>
  27 |   ) : <button  onClick={event => this.toggle(event)} > Show </button>}
  28 |"
`;

exports[`Django > svelte > Javascript Test > imports 1`] = `
"import Button from \\"./Button\\";

import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() disabled = false;

  render() {
    return (
      <div>
        <Button type=\\"button\\" disabled={this.disabled}>
          <Slot></Slot>
        </Button>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > lifecycleHooks 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  componentDidLoad() {
    console.log(\\"onMount\\");
  }
  disconnectedCallback() {
    console.log(\\"onDestroy\\");
  }
  componentDidUpdate() {
    console.log(\\"onAfterUpdate\\");
  }

  render() {
    return <div></div>;
  }
}
"
`;

exports[`Django > svelte > Javascript Test > reactive 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() name = \\"Steve\\";

  get lowercaseName() {
    return this.name.toLowerCase();
  }

  render() {
    return (
      <div>
        <input value={this.name} />
        Lowercase:
        {this.lowercaseName}
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > reactiveWithFn 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() a = 2;
  @State() b = 5;
  @State() result = null;

  calculateResult = function calculateResult(a_, b_) {
    this.result = a_ * b_;
  };

  componentDidUpdate() {
    this.calculateResult(this.a, this.b);
  }

  render() {
    return (
      <div>
        <input
          type=\\"number\\"
          onInput={(event) => (this.a = event.target.value)}
          value={this.a}
        />
        <input
          type=\\"number\\"
          onInput={(event) => (this.b = event.target.value)}
          value={this.b}
        />
        Result:
        {this.result}
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > slots 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  render() {
    return (
      <div>
        <Slot>default</Slot>
        <Slot name=\\"Test\\">
          <div>default</div>
        </Slot>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Javascript Test > style 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
  styles: \`
        input {
          color: red;
          font-size: 12px;
        }

        .form-input:focus {
          outline: 1px solid blue;
        }
\`,
})
export default class MyComponent {
  render() {
    return <input class=\\"form-input\\" />;
  }
}
"
`;

exports[`Django > svelte > Javascript Test > textExpressions 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() a = 5;
  @State() b = 12;

  render() {
    return (
      <div>
        normal:
        {this.a + this.b}
        <br />
        conditional
        {this.a > 2 ? \\"hello\\" : \\"bye\\"}
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > basic 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() name = \\"Steve\\";

  render() {
    return (
      <div>
        <input
          onInput={(event) => (this.name = event.target.value)}
          value={this.name}
        />
        Hello! I can run in React, Vue, Solid, or Liquid!
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > bindGroup 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() tortilla = \\"Plain\\";
  @State() fillings = [];

  render() {
    return (
      <div>
        <input
          type=\\"radio\\"
          value=\\"Plain\\"
          checked={this.tortilla === \\"Plain\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <input
          type=\\"radio\\"
          value=\\"Whole wheat\\"
          checked={this.tortilla === \\"Whole wheat\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <input
          type=\\"radio\\"
          value=\\"Spinach\\"
          checked={this.tortilla === \\"Spinach\\"}
          onInput={(event) => (this.tortilla = event.target.value)}
        />
        <br />
        <br />
        <input
          type=\\"checkbox\\"
          value=\\"Rice\\"
          checked={this.fillings === \\"Rice\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Beans\\"
          checked={this.fillings === \\"Beans\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Cheese\\"
          checked={this.fillings === \\"Cheese\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <input
          type=\\"checkbox\\"
          value=\\"Guac (extra)\\"
          checked={this.fillings === \\"Guac (extra)\\"}
          onInput={(event) => (this.fillings = event.target.value)}
        />
        <p>
          Tortilla:
          {this.tortilla}
        </p>
        <p>
          Fillings:
          {this.fillings}
        </p>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > bindProperty 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() value = \\"hello\\";

  render() {
    return <input value={this.value} />;
  }
}
"
`;

exports[`Django > svelte > Typescript Test > classDirective 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @Prop() disabled: any;

  @State() focus = true;

  render() {
    return (
      <input
        class={\`form-input \${props.disabled ? \\"disabled\\" : \\"\\"} \${
          state.focus ? \\"focus\\" : \\"\\"
        }\`}
      />
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > context 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() activeTab = 0;

  render() {
    return <div>{this.activeTab}</div>;
  }
}
"
`;

exports[`Django > svelte > Typescript Test > each 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() numbers = [\\"one\\", \\"two\\", \\"three\\"];

  render() {
    return (
      <ul>
        {this.numbers?.map((num) => (
          <li>{num}</li>
        ))}
      </ul>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > eventHandlers 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  log = function log(msg = \\"hello\\") {
    console.log(msg);
  };

  render() {
    return (
      <div>
        <button onClick={(a) => this.log(\\"hi\\")}>Log</button>
        <button onClick={(event) => this.log(event)}>Log</button>
        <button onClick={(event) => this.log(event)}>Log</button>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > html 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() html = \\"<b>bold</b>\\";

  render() {
    return <div innerHTML={this.html}></div>;
  }
}
"
`;

exports[`Django > svelte > Typescript Test > ifElse 1`] = `
"':' expected. (25:14)
  23 |       return (
  24 |
> 25 |         {this.show ? (
     |              ^
  26 |     <button  onClick={event => this.toggle(event)} > Hide </button>
  27 |   ) : <button  onClick={event => this.toggle(event)} > Show </button>}
  28 |"
`;

exports[`Django > svelte > Typescript Test > imports 1`] = `
"import Button from \\"./Button\\";

import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() disabled = false;

  render() {
    return (
      <div>
        <Button type=\\"button\\" disabled={this.disabled}>
          <Slot></Slot>
        </Button>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > lifecycleHooks 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  componentDidLoad() {
    console.log(\\"onMount\\");
  }
  disconnectedCallback() {
    console.log(\\"onDestroy\\");
  }
  componentDidUpdate() {
    console.log(\\"onAfterUpdate\\");
  }

  render() {
    return <div></div>;
  }
}
"
`;

exports[`Django > svelte > Typescript Test > reactive 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() name = \\"Steve\\";

  get lowercaseName() {
    return this.name.toLowerCase();
  }

  render() {
    return (
      <div>
        <input value={this.name} />
        Lowercase:
        {this.lowercaseName}
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > reactiveWithFn 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() a = 2;
  @State() b = 5;
  @State() result = null;

  calculateResult = function calculateResult(a_, b_) {
    this.result = a_ * b_;
  };

  componentDidUpdate() {
    this.calculateResult(this.a, this.b);
  }

  render() {
    return (
      <div>
        <input
          type=\\"number\\"
          onInput={(event) => (this.a = event.target.value)}
          value={this.a}
        />
        <input
          type=\\"number\\"
          onInput={(event) => (this.b = event.target.value)}
          value={this.b}
        />
        Result:
        {this.result}
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > slots 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  render() {
    return (
      <div>
        <Slot>default</Slot>
        <Slot name=\\"Test\\">
          <div>default</div>
        </Slot>
      </div>
    );
  }
}
"
`;

exports[`Django > svelte > Typescript Test > style 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
  styles: \`
        input {
          color: red;
          font-size: 12px;
        }

        .form-input:focus {
          outline: 1px solid blue;
        }
\`,
})
export default class MyComponent {
  render() {
    return <input class=\\"form-input\\" />;
  }
}
"
`;

exports[`Django > svelte > Typescript Test > textExpressions 1`] = `
"import { Component, Prop, h, State, Fragment } from \\"@django/core\\";

@Component({
  tag: \\"my-component\\",
})
export default class MyComponent {
  @State() a = 5;
  @State() b = 12;

  render() {
    return (
      <div>
        normal:
        {this.a + this.b}
        <br />
        conditional
        {this.a > 2 ? \\"hello\\" : \\"bye\\"}
      </div>
    );
  }
}
"
`;
